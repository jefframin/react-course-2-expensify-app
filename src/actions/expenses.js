import uuid from 'uuid'
import database from '../firebase/firebase'

export const addExpenseAction = ({ id, description, note, amount, created } = {}) => {
    return {
        type: 'ADD_EXPENSE',
        expense: {
        id: id,
        description: description,
        note: note,
        amount: amount,
        created: created
        }
    }
}

export const removeExpenseAction = ({ id } = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id: id
    }
}

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpenseAction({ id }))
        })
    };
}

export const editExpenseAction = (id, updates) => {
    return {
        type: 'EDIT_EXPENSE',
        id: id,
        updates: updates
    }
}

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note =  '',
            amount = 0,
            created = 0
        } = expenseData
        const expense = { description, note, amount, created }
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpenseAction({
                id: ref.key,
                ...expense
            }))
        })
    };
}

// set expenses
export const setExpenses = (expenses) => {
    return {
        type: 'SET_EXPENSES',
        expenses: expenses
    }
} 
        
export const startSetExpenses = () => {

    return (dispatch) => {

        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = []
            snapshot.forEach((element) => {
                expenses.push({
                    ...element.val(),
                    id: element.key
                });
            });
        dispatch(setExpenses(expenses));
    });
};
};