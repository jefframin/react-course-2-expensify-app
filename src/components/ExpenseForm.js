import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props)

    this.state ={
        description: props.expense ? props.expense.description : '',
        note: props.expense ? props.expense.note : '',
        amount: props.expense ? (props.expense.amount / 100).toString() : '',
        created: props.expense ? moment(props.expense.created) : moment(),
        calFocused: false,
        error: false
    }
    }

    onDescriptionChange = (event) => {
        const desc = event.target.value
        this.setState(() => {return {description: desc}})
    }

    onNoteChange = (event) => {
        const note = event.target.value
        this.setState(() => {return {note: note}})
    }

    onAmountChange = (event) => {
        const amount = event.target.value
        if (!amount || amount.match(/^[1-9]\d*(\.\d{0,2})?$/)) {
            this.setState(() => {return {amount: amount}})
        }
    }

    onDateChange = (created) => {
        if (created) {
          this.setState(() => {return {created: created}})
        }
    }

    onCalFocusChange = ({focused}) => {
        this.setState(() => {return {calFocused: focused}})
    }

    onSubmit = (event) => {
        event.preventDefault();
        if (!this.state.description || !this.state.amount) {
          this.setState(() => {return {error: true}})
        }
        else {
          this.setState(() => {return {error: false}})
          this.props.onSubmit({
              description: this.state.description,
              amount: parseFloat(this.state.amount, 10) * 100,
              created: this.state.created.valueOf(),
              note: this.state.note
          })
        }
    }

    render() {
        return (
            <div>
            {this.state.error && <p>please provide both description and amount</p>}
            <form onSubmit={this.onSubmit}>
            <input type="text" 
                   placeholder="Description" 
                   autoFocus
                   value={this.state.description}
                   onChange={this.onDescriptionChange}></input>
            <input type="text" 
                   value={this.state.amount}
                   placeholder="Amount"
                   onChange={this.onAmountChange}></input>
            <SingleDatePicker date={this.state.created}
                              onDateChange={this.onDateChange}
                              focused={this.state.calFocused}
                              numberOfMonths={1}
                              isOutsideRange={() => {return false}}
                              onFocusChange={this.onCalFocusChange}/>
            <textarea placeholder="Add a note for your expense (optional)"
                      value={this.state.note}
                      onChange={this.onNoteChange}></textarea>
            <button>Save Expense</button>
            </form>
            </div>
        );
    }
}