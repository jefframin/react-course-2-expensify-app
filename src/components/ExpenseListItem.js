import React from 'react'
import { Link } from 'react-router-dom';

const ExpenseListItem = ({id, description, amount, created}) => {
    return (
        <div>
          <h3>
          <Link to={`/edit/${id}`}>{description}</Link>
          </h3>
          <p>{amount} - {created}</p>
        </div>
    )
}

export default ExpenseListItem