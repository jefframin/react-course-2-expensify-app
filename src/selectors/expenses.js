import moment from 'moment'

const getVisibleExpenses = ( {expenses, filters} ) => {
    return expenses.filter((expense) => {
            const timeCreated = moment(expense.created)
            return (expense.description.toLowerCase().includes(filters.text.toLowerCase()) &&
                   (filters.startDate ? filters.startDate.isSameOrBefore(timeCreated, 'day') : true) &&
                   (filters.endDate ? filters.endDate.isSameOrAfter(timeCreated, 'day') : true))
        }).sort((a, b) => {
            switch (filters.order) {
                case 'amount':
                    return a.amount < b.amount ? -1 : 1
                case 'date':
                    return a.created < b.created ? -1 : 1
                default:
                    return 0
            }
        })
}

export default getVisibleExpenses