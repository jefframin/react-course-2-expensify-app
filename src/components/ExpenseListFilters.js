import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import { setTextFilterAction, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends React.Component {

    state = {
        calFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calFocused) => {
        this.setState(() => {
            return {calFocused: calFocused}
        })
    }

    onTextChange = (event) => {
       this.props.setTextFilter(event.target.value)
    }

    onSortChange = (event) => {
        event.target.value == 'amount' ? this.props.sortByAmount() : this.props.sortByDate()
    }

    render() {

    return (
        <div>
          <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
          <select value={this.props.filters.order} onChange={this.onSortChange} >
            <option value='date'>Date</option>
            <option value='amount'>Amount</option>
          </select>
          <DateRangePicker
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => {return false}}
                >
          </DateRangePicker>
        </div>
    )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilterAction(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)