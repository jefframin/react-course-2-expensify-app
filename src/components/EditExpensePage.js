import React from 'react';
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpenseAction } from '../actions/expenses'
import { removeExpenseAction } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit=(expense) => {
            this.props.editExpenseAction(this.props.expense.id, expense)
            this.props.history.push('/')
        }
    onClick=(event) => {
            this.props.removeExpenseAction({id: this.props.expense.id})
            this.props.history.push('/')
        }

    render() {
    return (
        <div>
          <ExpenseForm expense={this.props.expense}
                       onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
        </div>
    )
    }
}

/*
const EditExpensePage = (props) => {
    console.log(props)
    return (
        <div>
          <ExpenseForm expense={props.expense}
                        onSubmit={(expense) => {
            props.dispatch(editExpenseAction(props.expense.id, expense))
            props.history.push('/')
        }}/>
        <button onClick={(event) => {
            props.dispatch(removeExpenseAction({id: props.expense.id}))
            props.history.push('/')
        }}>Remove</button>
        </div>
    )
}
*/

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        }) 
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        editExpenseAction: (id, expense) => dispatch(editExpenseAction(id, expense)),
        removeExpenseAction: ({ id }) => dispatch(removeExpenseAction({ id }))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)