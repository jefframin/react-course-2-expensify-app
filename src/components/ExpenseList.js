import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => {
    return (
        <div>
        {
            props.expenses.length === 0 ? (
                <p>no expenses</p>
            ) : (
                props.expenses.map((expense) => {
                   return <ExpenseListItem key={expense.id} {...expense}/>
                }) 
            )
        }
        </div>

    )
}

const ConnectedExpenseList = connect((state) => {
    return {
        expenses: selectExpenses(state)
    }

})(ExpenseList)

export default ConnectedExpenseList