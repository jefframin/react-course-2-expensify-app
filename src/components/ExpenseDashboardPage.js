import React from 'react';
import ExpenseList from './ExpenseList'
import ExpensesSummary from './ExpensesSummary'
import ExpenseListFilters from './ExpenseListFilters'

const ExpenseDashboardPage = () => {
    return (
        <div>
        <ExpensesSummary />
        <ExpenseListFilters></ExpenseListFilters>
        <ExpenseList></ExpenseList>
        </div>
    )
}

export default ExpenseDashboardPage