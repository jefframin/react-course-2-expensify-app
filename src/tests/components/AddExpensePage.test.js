import React from 'react'
import { shallow } from 'enzyme'
import { AddExpensePage } from '../../components/AddExpensePage'
import expenses from '../fixtures/expenses'

let addExpenseSpy, history, wrapper;

beforeEach(() => {
    addExpenseSpy = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(<AddExpensePage startAddExpense={addExpenseSpy} history={history} />)
})

test('should render add expense page', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
    expect(history.push).toHaveBeenLastCalledWith('/dashboard')
    expect(addExpenseSpy).toHaveBeenLastCalledWith(expenses[2])
})