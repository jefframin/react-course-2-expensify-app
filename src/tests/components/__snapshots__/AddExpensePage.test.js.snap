// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense page 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="pag-header__title"
      >
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
